### PR TITLE
Paymill: Add currency and amount to store requests

### PR DESCRIPTION
**Pull request by jasonxp**
Paymill now requires a currency and amount to be specified for tokenization. These are used
for an authorization. We now set this authorization amount to $1 USD -- the minimum allowed -- in our ```store``` requests. 

The ```currency``` and ```money``` options  are translated to ```presentation.amount3D``` and ```presentation.currency3D``` in the request made to Paymill.

### CODE REVIEW

**Comment by curiousepic** [comment]
Can/should the amount be voided afterward? If so, we can use a `MultiResponse` to perform a void after the store.

**Comment by molbrown** [comment]
Was about to ask the same question ☝️

File: test/unit/gateways/paymill_test.rb [hunk 1]
```diff
@@ -181,6 +181,32 @@ def test_successful_store
     assert response.test?
   end
 
+  def test_store_includes_currency
+    expected_currency = 'USD'
+
+    @gateway.expects(:raw_ssl_request).with(
```
Comment (leila-alderman) [inline test/unit/gateways/paymill_test.rb]: Looking through the other unit tests in this file, none of them use `.with`. Why is it needed for this unit test?

File: test/unit/gateways/paymill_test.rb [hunk 2]
```diff
@@ -181,6 +181,32 @@ def test_successful_store
...
+      :get,
```
Comment (leila-alderman) [inline test/unit/gateways/paymill_test.rb]: What does this line do?

File: test/unit/gateways/paymill_test.rb [hunk 3]
```diff
@@ -181,6 +181,32 @@ def test_successful_store
...
+      store_endpoint(@credit_card, expected_currency, 100),
+      nil,
+      {}
+    ).returns(successful_store_response, successful_purchase_response)
```
Comment (leila-alderman) [inline test/unit/gateways/paymill_test.rb]: Since it's been determined that the `store` call no longer works without passing both `currency` and `amount`, what is the reason for creating two separate tests ensuring that these values are passed as opposed to just testing that the `store` calls are successful?

File: test/unit/gateways/paymill_test.rb [hunk 4]
```diff
@@ -181,6 +181,32 @@ def test_successful_store
...
+
+    @gateway.store(@credit_card)
+  end
+
+  def test_store_includes_amount
```
Comment (leila-alderman) [inline test/unit/gateways/paymill_test.rb]: Why have these set up as two separate tests? It looks like both of these tests are passing in both `currency` and `amount`, so I'm not seeing the difference between these tests.

File: test/unit/gateways/paymill_test.rb [hunk 5]
```diff
@@ -225,6 +251,10 @@ def test_transcript_scrubbing
 
   private
 
+  def store_endpoint(credit_card, currency, amount)
```
Comment (leila-alderman) [inline test/unit/gateways/paymill_test.rb]: Why is this endpoint needed? It looks like there were already unit tests set up to test the `store` action, so why is this needed now?

File: lib/active_merchant/billing/gateways/paymill.rb [hunk 6]
```diff
@@ -48,6 +48,11 @@ def void(authorization, options={})
       end
 
       def store(credit_card, options={})
+        # The store request requires a currency and amount of at least $1 USD.
+        # This is used for an authorization.
```
Comment (leila-alderman) [inline lib/active_merchant/billing/gateways/paymill.rb]: I don't understand why this comment is talking about authorization within the `store` method. Is the issue that a `store` transaction for PayMill now runs an authorization first before storing the card? Do all `store` methods work that way?

**Comment by jasonxp** [comment]
@curiousepic and @molbrown My understanding is that the "store" authorization is something that Paymill does internally. It doesn't appear to be possible to void it from our end, since we don't receive an authorization ID back in response. This is what the store response looks like: https://github.com/activemerchant/active_merchant/blob/master/test/unit/gateways/paymill_test.rb#L229.

File: lib/active_merchant/billing/gateways/paymill.rb (hunk 6 above)
Comment (jasonxp) [reply lib/active_merchant/billing/gateways/paymill.rb]: Yep, that is the issue. The ```store``` requests for Paymill started resulting in errors when Paymill added ```currency``` and ```amount``` as required fields for the ```store``` operation. I can tweak the comment to be more explicit in saying that the authorization is something Paymill does internally. I'm not aware of any other gateway ```store```s that work this way.

File: test/unit/gateways/paymill_test.rb (hunk 5 above)
Comment (jasonxp) [reply test/unit/gateways/paymill_test.rb]: This is a method that generates the endpoint URL that the ```store``` operation makes its request to. It's used by the new tests added here. I named the method as a noun for consistency with the existing helper methods like ```successful_store_response```, but it could also be named something like ```generate_store_endpoint```. I'll go with ```store_endpoint_url``` since that seems like it would be more clear.

File: test/unit/gateways/paymill_test.rb (hunk 1 above)
Comment (jasonxp) [reply test/unit/gateways/paymill_test.rb]: The purpose of the test is to ensure that the currency gets included in the request sent to Paymill. So, the expectation is that when we call the method to make an SSL request, it's done ```with``` the currency parameter set to the right value.

File: test/unit/gateways/paymill_test.rb (hunk 2 above)
Comment (jasonxp) [reply test/unit/gateways/paymill_test.rb]: ```store_endpoint``` is just a method that returns a string. That string is the second expected argument in the ```raw_ssl_request``` call.

Comment (leila-alderman) [reply test/unit/gateways/paymill_test.rb]: Oooh, I think I see now. So the `.with()`method requires _all_ of the parameters that are passed to the given function, meaning that you needed to pass in everything here:
```
def raw_ssl_request(method, endpoint, data, headers = {})
```
https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/posts_data.rb#L47

That's where I had been confused. I didn't know what the four different parameters listed out here were doing.

File: test/unit/gateways/paymill_test.rb (hunk 1 above)
Comment (leila-alderman) [reply test/unit/gateways/paymill_test.rb]: Okay, that makes sense. Why was this check added as a unit test but not as a remote test?

File: test/unit/gateways/paymill_test.rb (hunk 5 above)
Comment (leila-alderman) [reply test/unit/gateways/paymill_test.rb]: Nice, thanks!

**Comment by curiousepic** [comment]
Oh duh, it likely doesn't actually capture, and they just let the auth timeout. 👍

**Comment by molbrown** [comment]
I do wonder why they ask for an amount though, if that is the case. You would think they could just do an internal $1 auth on any store request. I am skeptical that this could be for a purchase+store action, and I find their docs really challenging to navigate.

Do we have confirmation from Paymill/ other confidence that the store amount does what we think it does?

File: test/unit/gateways/paymill_test.rb (hunk 1 above)
Comment (jasonxp) [reply test/unit/gateways/paymill_test.rb]: Good question. When Paymill responds to a ```store``` request, it doesn't return anything different in the case of a currency and amount being sent vs. not being sent. So, there's nothing that we can check in a remote test. The options are being set in the code that implements the request, but a user of that code cannot see those details. The remote test is performing its testing as if it were a user of the code, only calling the public methods (e.g., ```gateway.store```) and checking the return values. The more general theme here is unit testing vs. integration testing.

Comment (leila-alderman) [reply test/unit/gateways/paymill_test.rb]: I remember that there was a similar issue when trying to test the new `tx_source` GSF for the MiGS gateway. Would replicating the approach I took for those remote make sense here?
https://github.com/activemerchant/active_merchant/commit/4493b55b5bf49f854bcd43e08cae1b9e30b62eb2#diff-a1a05cf6a8356f8c6539f34382a90bf6R110

Comment (jasonxp) [reply test/unit/gateways/paymill_test.rb]: That is indeed a similar case. I don't think a test like that is possible in this case, though, because the options being passed are hardcoded in the ```store``` implementation. There isn't a way for the caller (i.e. our remote test) to change them to something invalid. FWIW, I did manually test that.
